Mathlib/Tactic/NthRewrite.lean

```diff
@@ -17,11 +17,48 @@ namespace Mathlib.Tactic
 
 open Lean Elab Tactic Meta Parser.Tactic
 
-/-- `nth_rewrite` is a variant of `rewrite` that only changes the nth occurrence of the expression
-to be rewritten.
+/-- `nth_rewrite` is a variant of `rewrite` that only changes the `n`ᵗʰ _occurrence_ of the
+expression to be rewritten. `nth_rewrite n [eq₁, eq₂, ..., eqₘ]` will rewrite the `n`ᵗʰ _occurrence_
+of each of the `m` equalities `eqᵢ`in that order. Occurrences are counted beginning with `1`.
+Occurrences are counted beginning with `1`. For example,
+```lean
+example (h : a = 1) : a + a + a + a + a = 5 := by
+  nth_rewrite 2 [h]
+/-
+a: ℕ
+h: a = 1
+⊢ a + 1 + a + a + a = 5
+-/
+```
+Notice that the second occurrence of `a` from the left has been rewritten by `nth_rewrite`.
+
+If a term `t` is introduced by rewriting with `eqᵢ`, then this instance of `t` will be counted
+as an _occurrence_ of `t` for all subsequent rewrites of `t` with `eqⱼ` for `j > i`. This behaviour
+is illustrated by the example below
+```lean
+example (h : a = a + b) : a + a + a + a + a = 0 := by
+  nth_rewrite 3 [h, h]
+/-
+a b: ℕ
+h: a = a + b
+⊢ a + a + (a + b + b) + a + a = 0
+-/
+```
+In this example, the first `nth_rewrite` with `h` introduces an additional occurrence of `a` in
+the goal. That is, the goal state after the first rewrite looks like below
+```lean
+/-
+a b: ℕ
+h: a = a + b
+⊢ a + a + (a + b) + a + a = 0
+-/
+```
+This new instance of `a` also turns out to be the third _occurrence_ of `a`.  Therefore,
+the next `nth_rewrite` with `h` rewrites this `a`.
```

Comment:
This part I find quite hard to read, and I do find this behaviour intuitive as is, not requiring much explanation.

Maybe this entire paragraph could be shortened to something like the following?

> `nth_rewrite` with multiple arguments recalculates the occurences after each step newly, so for `(h : a = a + b)`, `nth_rewrite 3 [h, h]` changes `a + a + a + a + a = 0` to `a + a + (a + b) + a + a = 0` in the first step and then to `a + a + ((a + b) + b) + a + a = 0` in the second.

Reply:
In a previous iteration I had such a short sentence. The problem is if someone is not comfortable with binder substitution and name capture issues, then it is a bit hard for them to see what is happening without a detailed example.

Reply:
Fair enough. I also think the new iteration with the goal states is good.